### DISCUSSION TITLE
How to modify nodes selected by FHIRPath "Select" function

### DISCUSSION THREAD

**Discussion by BoyaWu10** [FirelyTeam/firely-net-sdk#1817] [Q&A] [answered]
Hi, 

We have a question about `IEnumerable<ITypedElement> Select(this ITypedElement input, string expression, EvaluationContext ctx = null)`.

After parsing the input FHIR resource to `ITypedElement`:
```csharp
var provider = new PocoStructureDefinitionSummaryProvider();
var rootNode = FhirJsonNode.Parse(inputFileContent).ToTypedElement(provider);
```
we could select the target node by FHIRPath using above function:
```csharp
var matchNode = rootNode.Select("Patient.identifier").FirstOrDefault();
```
Matched node with instance type `ScopedNode` would be returned.

If we want to modify this node using `Add/Remove/Replace` functions in `ElementNode`, is there any method to convert this `ScopedNode` to `ElementNode` safely?

Can we achieve this by converting the input to `ElementNode` before `Select`, and then get it from `ScopedNode.Current`?
```csharp
// Convert to ElementNode
var rootNode = ElementNode.FromElement(FhirJsonNode.Parse(inputFileContent).ToTypedElement(provider));
var matchNode = rootNode.Select("Patient.identifier").FirstOrDefault();
if (matchNode != null)
{
    // Will this always work?
    var elementNode = (ElementNode)((ScopedNode)matchNode).Current;
}
```

Thanks.

**Comment by ewoutkramer** [comment, marked as answer]
Yes, that is the approach I would take as well.  Navigating a ScopedNode using FhirPath will generally just call `Children()` on the ScopedNode, which will do this:

``` csharp
        public IEnumerable<ITypedElement> Children(string? name = null) =>
            Current.Children(name).Select(c => new ScopedNode(this, ParentResource, c, _fullUrl));
```

This fragment shows that while navigating down the tree using FhirPath, new `ScopedNode`s are created using a private constructor, which just wraps the `ElementNode` children (`c`) of the current `ElementNode`, and they will be exposed through `Current` in the child nodes.

So this should work!

**Reply by BoyaWu10** [reply]
Thanks so much, Ewout! And why do we choose to wrap FhirPath input in `ScopedNode` before evaluating (https://github.com/FirelyTeam/firely-net-common/pull/116)? When should we consider to use `ScopedNode` instead of other nodes? In previous versions of the library, if we pass an `ElementNode` to function `Select()`, it would return `ElementNode` as well.

**Reply by ewoutkramer** [reply]
This is because some functions in FhirPath require awareness of more context than ITypedElement is capable of providing.  For example, ScopedNode keeps track of the contained resources that are encountered while walking the tree, so that FhirPath's `resolve()` function is able to locate contained resources.  The same holds for the FhirPath variables `%resource` and `%rootResource`.   ScopedNode allows code to walk up in the hierarchy so these variables can be determined.

**Reply by BoyaWu10** [reply]
I see. Thanks so much for your kind help!
